### PR TITLE
Add shape anisotropy analysis to cluster move

### PR DESCRIPTION
**Pull request by mlund**
Sample relative cluster size anisotropy during cluster move propagation. This is the same algorithm as used in `polymershape`. Ping @MarcoPolimeni

### CODE REVIEW

File: src/clustermove.cpp [hunk 1]
```diff
@@ -2,61 +2,96 @@
...
+/**
+ * Search for molecules participating in the cluster move.
+ * This should be called for every move event as a
+ * grand canonical move may have changed the number
+ * of particles.
+ *
+ * Atomic groups and inactive groups are ignored.
+ */
+void FindCluster::updateMoleculeIndex() {
+    auto to_group_index = [&](const Tgroup &group) { return &group - &spc.groups.front(); };
```
Comment (rc83) [inline src/clustermove.cpp:75]: If possible, hiding the pointer arithmetic (black magic) in a named function at groups or similar would be nice. ;)

**Review by rc83 — COMMENTED** [review]
Nice extension and refactorization! I like breaking the code to smaller and well described pieces very much. BTW Shall we begin to slowly introduce concepts? ;) Mainly nitpicking comments about adhering to the code style follows.

Only one conceptually bigger issue: Should not be the old `Average` template for `double` a specialized template of `AverageObj`, renamed to plain `Average`? It should not take much effort.

Because the commit history looks a bit messy, I suggest to merge it by myself after you give the final touch to the code.

File: src/average.h [hunk 2]
```diff
@@ -85,5 +85,66 @@ namespace Faunus
...
+       * @brief Simple class to average data contained in objects
+       * @tparam T Type to average
+       * @tparam int_t Unsigned interger type
+       *
+       * It is required that `T` has the following operator overloads:
+       * - `T& operator+=(const &T)`
+       * - `T operator*(double) const`
+       */
+#if __cplusplus > 201703L
+  template <Averageable T, typename int_t = unsigned long int> class AverageObj {
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/average.h:105]: I agree. There are some remaining issues in that to replace `Avererage` a `pow()` function is needed to rms and stdev. I have added a bit more in this direction but not done.

File: src/average.h [hunk 3]
```diff
@@ -85,5 +85,66 @@ namespace Faunus
       } // de-serialize from stream
   };
 
+#if __cplusplus > 201703L
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/average.h:88]: please do!

File: src/average.h [hunk 4]
```diff
@@ -85,5 +85,66 @@ namespace Faunus
...
+      T sum; // make sure constructors zero this!
+    public:
+      AverageObj() : sum(T()){}; //!< Construct from empty object
+
+      AverageObj(const T &value) : number_of_samples(1), sum(value){};
+
+      //! Add to average
+      AverageObj &operator+=(const T &value) {
+          if (number_of_samples == std::numeric_limits<int_t>::max()) {
+              throw std::runtime_error("Integer overflow - maximum samples reached");
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/average.h:120]: done

File: src/clustermove.h [hunk 5]
```diff
@@ -2,52 +2,135 @@
...
+    std::unique_ptr<std::ostream> stream; //!< log current cluster into (compressed) stream
+    std::map<size_t, size_t> size_distribution;                                  //!< distribution of cluster sizes
+    std::map<size_t, AverageObj<Geometry::ShapeDescriptors>> shape_distribution; //!< average shape
+    std::map<size_t, std::ofstream> pqr_distribution;                            //!< Each cluster size is saved to disk
+    decltype(pqr_distribution)::iterator findPQRstream(size_t cluster_size);     //!< Create or find PQR files stream
+
+    /** brief Calculates the gyration tensor for a collection of groups */
+    template <typename Range>
+    Tensor gyrationGroups(
+        const Range &groups, const Point &COM, const Geometry::BoundaryFunction boundary = [](auto &) {}) {
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/clustermove.h:27]: done

File: src/geometry.h [hunk 6]
```diff
@@ -619,6 +619,64 @@ Tensor gyration(
...
+Tensor gyration(
+    position_iterator begin, position_iterator end, mass_iterator mass, const Point &mass_center,
+    const BoundaryFunction boundary = [](auto &) {}) {
+    Tensor S = Tensor::Zero();
+    double total_mass = 0.0;
+    std::for_each(begin, end, [&](const auto &position) {
+        Point r = position - mass_center; // get rid...
+        boundary(r);                      // ...of PBC (if any)
+        S += *mass * r * r.transpose();
+        total_mass += *mass;
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/geometry.h:651]: done

File: src/clustermove.cpp (hunk 1 above)
Comment (mlund) [reply src/clustermove.cpp:75]: Subtracting two addresses doesn't seem so magical to me ;-)

File: src/average.h [hunk 7]
```diff
@@ -85,5 +85,66 @@ namespace Faunus
...
+              number_of_samples++;
+          }
+          return *this;
+      }
+      //! Calculate average
+      T avg() const {
+          if (number_of_samples > 0) {
+              return sum * (1.0 / static_cast<double>(number_of_samples));
+          } else {
+              return sum;
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/average.h:132]: I agree; average requires an overhaul

File: src/clustermove.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlund) [reply src/clustermove.h:27]: done

**Comment by mlund** [comment]
@rc83: I pushed a new version addressing the comments. Feel free to mangle/merge as you see fit!

**Review by rc83 — APPROVED** [review]

**Pull request merged**